### ISSUE TITLE
[Bug?]: Invalid typescript types resolved with yarn 3

### ISSUE THREAD

**Issue by Js-Brecht** [yarnpkg/berry#3299]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

There seems to be some issue with resolving typescript types for a package when importing modules from subdirectories/exports.

A couple screenshots that might help explain:
* The error displayed in vscode
![image](https://user-images.githubusercontent.com/1935258/129504947-3cf84e1f-1104-46b4-9123-c740693d7875.png)
* The type selected when I use `Go To Definition`
![image](https://user-images.githubusercontent.com/1935258/129505001-9c8ac631-5367-4186-8a00-eb77ba1fe4ef.png)

If I import `get` first, then the types for `set` will point to `get`.

Perhaps this has to do with `arcanis.vscode-zipfs`, because I remember that you didn't used to be able to use `Go To Definition` when using yarn 2.

The problem does not exist when I use Typescript with a more common package manager.

### To reproduce

Reproduction repository: https://github.com/Js-Brecht/yarn3-tsc-type-resolution-issue

Should be able to run `yarn` from the repository, and then try running `yarn tsc`... it throws an error where it shouldn't.

![image](https://user-images.githubusercontent.com/1935258/129505320-386c6534-77e1-4b96-b72a-3c9cf7888b1c.png)

### Environment

```
  System:
    OS: Linux 5.11 Ubuntu 21.04 (Hirsute Hippo)
    CPU: (8) x64 Intel(R) Core(TM) i7-2600K CPU @ 3.40GHz
  Binaries:
    Node: 12.21.0 - /tmp/xfs-43d13f8b/node
    Yarn: 3.0.1 - /tmp/xfs-43d13f8b/yarn
    npm: 6.14.11 - ~/.nvm/versions/node/v12.21.0/bin/npm
```

**Comment by merceyz** [comment]
Duplicate of https://github.com/yarnpkg/berry/issues/3058

**Comment by merceyz** [comment]
It's a bug in the PnP patch for TypeScript, you can disable PnP and go back to `node_modules` by running
```sh
yarn config set nodeLinker node-modules && yarn
```
or you can try the fix in https://github.com/yarnpkg/berry/pull/3297 by running
```sh
yarn set version from sources --branch 3297 && yarn
```

**Comment by Js-Brecht** [comment]
> Duplicate of #3058

Oh, my bad.  I tried to find an existing ticket, but was looking for the wrong thing :laughing:.

> or you can try the fix in #3297 by running

That works.  I get an initial error when using `Go To Definition`, but I think it has to do with the zipfs (nothing to do with that update, either).

```
Unable to open 'object.d.ts': Unexpected model for TextResourceEditorInput: zip:/home/jeremy/dev/test/yarn3-erroneous-types/.yarn/cache/%40types-lodash-npm-4.14.172-e52063c461-5fc51ea40d.zip/node_modules/%40types/lodash/common/object.d.ts.
```

It works on the second try.  I can open a new ticket for that; just don't know if it should be on this repo, or somewhere else?  I can reproduce it pretty reliably.

**Comment by WoodyWoodsta** [comment]
@Js-Brecht I'm getting this error fairly consistently. It would be great if you could open an issue.
